internal/bufferpool/bufferpool.go

```diff
@@ -23,24 +23,68 @@
 package bufferpool
 
 import (
-	"bytes"
+	"flag"
 	"sync"
 )
 
-var _bufferPool = sync.Pool{
-	New: func() interface{} {
-		return &bytes.Buffer{}
-	},
+var _pool = NewPool()
+
+// Option configures a buffer pool.
+type Option func(*Pool)
+
+// Pool represents a buffer pool with a set of options.
+type Pool struct {
+	testDetectUseAfterFree bool
+	pool                   sync.Pool
 }
 
-// Get returns a new Buffer from the Buffer pool that has been reset.
-func Get() *bytes.Buffer {
-	buf := _bufferPool.Get().(*bytes.Buffer)
-	buf.Reset()
+func init() {
+	// FIXME: This is a huge hack just for prototyping.
+	// This detects whether we're running as part of a test, and if so
+	// enables the use-after-free detection.
+	if flag.Lookup("test.v") != nil {
+		_pool = NewPool(DetectUseAfterFreeForTests())
+	}
+}
+
+// NewPool returns a pool that we can allocate buffers from.
+func NewPool(opts ...Option) *Pool {
+	pool := &Pool{}
+	for _, opt := range opts {
+		opt(pool)
+	}
+	return pool
+}
+
+// DetectUseAfterFreeForTests is an option that allows unit tests to detect
+// bad use of a pooled buffer after it has been released to the pool.
+func DetectUseAfterFreeForTests() Option {
```

Comment:
do these functions need to be public?  (it's all internal, so it's NBD, but, people probably shouldn't be calling the NewPool and Option function outside of the package).

Reply:
In future, we'll probably want to support using the use-after-free detection outside of unit tests (e.g., in demo/test services) for validation.

For that, we'll need to export options and provide some way for users to configure it.